cpp/src/arrow/compute/kernels/scalar_cumulative_sum.cc

```diff
@@ -0,0 +1,147 @@
+          arrow::internal::CopyBitmap(input->null_bitmap_data(), input->offset(),
+                                      input->length(), out_bitmap, out_offset);
+          output->null_count = input->null_count();
+        } else {
+          bit_util::SetBitsTo(out_bitmap, out_offset, input->length(), true);
+          output->null_count = 0;
+        }
+
+        CumulativeSum(input, output, start);
+        return Status::OK();
+      case Datum::CHUNKED_ARRAY:
+        const auto& input = batch[0].chunked_array();
+
+        ArrayVector out_chunks;
+        for (const auto& chunk : input->chunks()) {
+          auto out_chunk = std::make_shared<ArrayData>(
+              chunk->type(), chunk->length(), chunk->null_count(), chunk->offset());
+
+          uint8_t* out_chunk_bitmap = out_chunk->buffers[0]->mutable_data();
+          if (chunk->data()->MayHaveNulls()) {
+            arrow::internal::CopyBitmap(chunk->null_bitmap_data(), chunk->offset(),
+                                        chunk->length(), out_chunk_bitmap,
+                                        out_chunk->offset());
+            out_chunk->null_count = chunk->null_count();
+          } else {
+            bit_util::SetBitsTo(out_chunk_bitmap, out_chunk->offset(), chunk->length(),
+                                true);
+            out_chunk->null_count = 0;
+          }
+
+          CType last_value = CumulativeSum(chunk, out_chunk, start);
+          start = last_value;
+          out_chunks.push_back(MakeArray(std::move(out_chunk)));
+        }
+
+        *out->chunked_array() = ChunkedArray(out_chunks, input->type());
+        return Status::OK();
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'cumulative_sum': ",
+            batch[0].ToString());
+    }
+  }
+
+  static std::shared_ptr<KernelSignature> GetSignature(detail::GetTypeId get_id) {
+    return KernelSignature::Make(
+        {InputType::Array(get_id.id), InputType::Scalar(get_id.id)},
+        OutputType(FirstType));
+  }
+};
+
+const FunctionDoc cumulative_sum_doc(
+    "Compute the cumulative sum over an array of numbers",
+    ("`values` must be an array of numeric type values.\n"
+     "`start` is a single value of the same type.\n"
+     "Return an array which is the cumulative sum computed over `values`\n"
+     "`start` is an optional starting sum of computation."),
+    {"values", "start"});
+
+void RegisterVectorCumulativeSum(FunctionRegistry* registry) {
+  auto add_kernel = [&](detail::GetTypeId get_id, ArrayKernelExec exec,
+                        std::shared_ptr<ScalarFunction> func) {
+    ScalarKernel kernel;
+    kernel.mem_allocation = MemAllocation::type::PREALLOCATE;
+    kernel.signature = CumulativeSumFunctor<NumberType>::GetSignature(get_id.id);
+    kernel.exec = std::move(exec);
+    DCHECK_OK(func->AddKernel(std::move(kernel)));
+  };
+
+  auto cumulative_sum = std::make_shared<ScalarFunction>(
```

Comment:
cumulative sum should be implemented as a [`VectorFunction`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/function.h#L177-L180), not `ScalarFunction`.